lib/core.mli

```diff
@@ -99,233 +99,230 @@ val clone_element : element -> 'props -> element
-
-val use_effect1 : (unit -> (unit -> unit) option_undefined) -> 'a array -> unit
-  [@@js.custom
-    val use_effect1_internal :
+    val _use_layout_effect :
          Imports.react
       -> (unit -> (unit -> unit) option_undefined)
       -> 'a array
       -> unit
-      [@@js.call "useEffect"]
+      [@@js.call "useLayoutEffect"]
 
-    let use_effect1 callback watchlist =
-      use_effect1_internal Imports.react callback watchlist]
+    let use_effect_once ?(before_render = false) f =
+      (if before_render then _use_layout_effect else _use_effect)
+        Imports.react f [||]]
 
-val use_effect2 : (unit -> (unit -> unit) option_undefined) -> 'a * 'b -> unit
+val use_effect_on_change :
+     ?before_render:bool
+  -> (unit -> (unit -> unit) option_undefined)
+  -> 'a array
+  -> unit
   [@@js.custom
-    val use_effect2_internal :
+    val _use_effect :
          Imports.react
       -> (unit -> (unit -> unit) option_undefined)
-      -> 'a * 'b
+      -> 'a array
       -> unit
       [@@js.call "useEffect"]
 
-    let use_effect2 callback watchlist =
-      use_effect2_internal Imports.react callback watchlist]
-
-val use_effect3 :
-  (unit -> (unit -> unit) option_undefined) -> 'a * 'b * 'c -> unit
-  [@@js.custom
-    val use_effect3_internal :
+    val _use_layout_effect :
          Imports.react
       -> (unit -> (unit -> unit) option_undefined)
-      -> 'a * 'b * 'c
+      -> 'a array
       -> unit
-      [@@js.call "useEffect"]
+      [@@js.call "useLayoutEffect"]
 
-    let use_effect3 callback watchlist =
-      use_effect3_internal Imports.react callback watchlist]
+    let use_effect_on_change ?(before_render = false) f values =
+      (if before_render then _use_layout_effect else _use_effect)
+        Imports.react f values]
 
-val use_effect4 :
-  (unit -> (unit -> unit) option_undefined) -> 'a * 'b * 'c * 'd -> unit
+val use_effect_on_change1 :
+  ?before_render:bool -> (unit -> (unit -> unit) option_undefined) -> 'a -> unit
   [@@js.custom
-    val use_effect4_internal :
+    val _use_effect :
          Imports.react
       -> (unit -> (unit -> unit) option_undefined)
-      -> 'a * 'b * 'c * 'd
+      -> 'a array
       -> unit
       [@@js.call "useEffect"]
 
-    let use_effect4 callback watchlist =
-      use_effect4_internal Imports.react callback watchlist]
-
-val use_effect5 :
-  (unit -> (unit -> unit) option_undefined) -> 'a * 'b * 'c * 'd * 'e -> unit
-  [@@js.custom
-    val use_effect5_internal :
+    val _use_layout_effect :
          Imports.react
       -> (unit -> (unit -> unit) option_undefined)
-      -> 'a * 'b * 'c * 'd * 'e
+      -> 'a array
       -> unit
-      [@@js.call "useEffect"]
+      [@@js.call "useLayoutEffect"]
 
-    let use_effect5 callback watchlist =
-      use_effect5_internal Imports.react callback watchlist]
+    let use_effect_on_change1 ?(before_render = false) f a =
+      (if before_render then _use_layout_effect else _use_effect)
+        Imports.react f [|a|]]
```

Comment:
This is 100%opinion, but I think this change is going to lead to a lot of confusion as diverges from rescript-react, as was seen with the tests endless rendering. Why not leave the array wrapping outside the function?

Reply:
It's leaking internal implementation details. There's no reason the user should need to know that this uses an array internally, and have to wrap it manually. It's also inconsistent with all the other `use_effect` functions, which all take a tuple instead of an array. Because, of course, tuples are represented as arrays in JS. But there is no 1-tuple that maps to a single-element array, because that makes absolutely no sense from an OCaml perspective. Would you consider an API like this reasonable if implemented in pure OCaml? If not, why should we expect user's to find it reasonable just because it binds to JavaScript?

I think the reason `rescript-react` has chosen this design is solely because they obsess over zero-cost bindings, irregardless of cognitive cost. If you interact very closely with JavaScript and expect your users to deeply understand this model, it might make sense, but I'm not sure that transfers all that well to jsoo.

You do point out a real problem though, and it's a hard one to fix without continuing the broken model of `rescript-react`...